BC-6009 Добавил обработчик достижений + абстрактный обработчик

Dregid opened the pull request:
Тесты буду реализованы в отдельной таске.

По какой то причине jackson не может преобразовать json в наш ивент. Вылезает MismatchedInputException, так и не смогли с Расулом разобраться в чем беда.

Dregid committed:
Implemented an abstract handler
- Created EventHandler interface
- Created a handler for the achievement "Sensei"
- Made actual changes to the trainees

mighty-mhsl commented on src/main/java/faang/school/achievement/handler/SenseyAchievementHandler.java:
```diff
@@ -0,0 +1,50 @@
+
+    @Override
+    protected boolean hasAchievement(MentorshipStartEvent event, Achievement achievement) {
+        return !achievementService.hasAchievement(event.getMentorId(), achievement.getId());
```
У нас метод называется hasAchievement, но на самом деле проверяет, что у юзера нет достижения

mighty-mhsl commented on src/main/java/faang/school/achievement/handler/AbstractAchievementHandler.java:
```diff
@@ -0,0 +1,44 @@
+
+    private void hasComplete(Achievement achievement, AchievementProgress achievementProgress, T event) {
+        if (achievement.getPoints() <= achievementProgress.getCurrentPoints()) {
+            achievementService.saveAchievement(getUserAchievement(achievement, event));
```
Что если у нас требовалось 100 очков для получения достижения, а у нас сейчас 99 
Пришли одновременно 2 эвента
Первый проверил, что нет у юзера достижения
Второй проверил, что нет достижения
Первый увеличил счетчик на 1
Второй увеличил счетчик на 1
Первый пытается сохранить новое достижение юзеру
Второй пытается сохранить новое достижение юзеру

Что будет?

Dregid replied:
Страх и ненависть. Как минимум конфликт со стороны бд будет. 
Был вариант использовать вторую проверку на наличие достижения. Но я подумал что два потока могут одновременно пройти эту проверку. 
А как грамотно использовать оптимистичный лок я так и не понял, Влад АИ вообще не смог контекста понять (хотя я его криво подал)
Бот предлагал JTA использовать для контроля транзакций, но мне казалось что проблему можно решить иначе

mighty-mhsl commented on src/main/java/faang/school/achievement/handler/AbstractAchievementHandler.java:
```diff
@@ -0,0 +1,44 @@
+    protected abstract boolean hasAchievement(T event, Achievement achievement);
+    protected abstract void createProgressIfNecessary(T event, Achievement achievement);
+    protected abstract AchievementProgress getAchievementProgress(T event, Achievement achievement);
+    protected abstract UserAchievement getUserAchievement(Achievement achievement, T event);
```
Нужно ли нам во все эти методы передавать целый generic эвент, если по факту нас интересует в этом эвенте только id юзера, для которого мы что-то ищем? Может, нам просто передавать id, и тогда мы реализацию этих методов тоже сможем в этот класс занести, т.к. она будет идентичной для любых юзеров вне зависимости от конкретного эвента

Dregid replied:
И вправду. Понял, исправляйкенс

mighty-mhsl commented on src/main/java/faang/school/achievement/handler/SenseyAchievementHandler.java:
```diff
@@ -0,0 +1,50 @@
+
+    @Override
+    public void handle(MentorshipStartEvent event) {
+        super.handle(event);
```
Можно не переопределять, мы же наследуем этот метод

Dregid replied:
А вот это уже а вот это уже вот

mighty-mhsl commented on src/main/java/faang/school/achievement/handler/SenseyAchievementHandler.java:
```diff
@@ -0,0 +1,50 @@
+
+    @Override
+    protected Achievement getAchievement() {
+        return super.getAchievement();
```
Тоже можно не переопределять

mighty-mhsl commented on src/main/java/faang/school/achievement/handler/AbstractAchievementHandler.java:
```diff
@@ -0,0 +1,44 @@
+        if (hasAchievement(event, achievement)) {
+            createProgressIfNecessary(event, achievement);
+            AchievementProgress achievementProgress = getAchievementProgress(event, achievement);
+            achievementProgress.increment();
```
Что если:
У нас счетчик 99
Приходят 2 эвента одновременно
Первый получает сущность из БД со счетчиком 99
Второй получает ту же сущность из БД
Первый увеличивает счетчик до 100
Второй увеличивает счетчик до 100

Мы потеряли очко достижений

Dregid replied:
Думали об этом. Но проблему разных клонов сервиса меня сбила и я так и не нашел ответ на этот вопрос

mighty-mhsl commented on src/main/java/faang/school/achievement/messaging/listener/AbstractEventListener.java:
```diff
@@ -0,0 +1,27 @@
+@RequiredArgsConstructor
+public abstract class AbstractEventListener<T> {
+    protected final ObjectMapper objectMapper;
+    protected final EventHandler<T> eventHandler;
```
Много должно быть. На каждое событие может быть несколько обработчиков определенного типа. Они должны проверять, подходят ли они под данный тип эвента. Те, которые подходят, обрабатывают его

Dregid replied:
Эм, идею не понял

mighty-mhsl commented on src/main/java/faang/school/achievement/messaging/listener/AbstractEventListener.java:
```diff
@@ -0,0 +1,27 @@
+    protected final ObjectMapper objectMapper;
+    protected final EventHandler<T> eventHandler;
+
+    protected T eventMapper(Message message, Class<T> eventType) {
```
Методы глаголами называем

Dregid replied:
В смысле, что за подстава :D Во всех сервисах где есть слушатели такой же метод)

Dregid committed:
Handler Refactoring

Dregid committed:
Переписан слушатель
- Исправлены баги в конфиге
- обновлены таблицы
- внедрен новый модуль для jackson

Dregid committed:
Отрефачен маппер

Dregid committed:
Merge branch 'unicorn-master' into feature-BC-6009

# Conflicts:
#	src/main/java/faang/school/achievement/config/RedisConfig.java
